influxdb3_server/src/http.rs

```diff
@@ -1491,70 +1493,173 @@ async fn record_batch_stream_to_body(
-        let mut writer =
-            TrackedMemoryArrowWriter::try_new(&mut bytes, batches[0].schema(), mem_pool)?;
-        for batch in batches {
-            writer.write(batch)?;
-        }
-        writer.close()?;
-        Ok(Bytes::from(bytes))
-    }
-
     match format {
         QueryFormat::Pretty => {
             let batches = stream.try_collect::<Vec<RecordBatch>>().await?;
-            to_pretty(batches).map(Body::from)
+            Ok(Body::from(Bytes::from(format!(
+                "{}",
+                pretty::pretty_format_batches(&batches)?
+            ))))
         }
         QueryFormat::Parquet => {
-            let batches = stream.try_collect::<Vec<RecordBatch>>().await?;
-            to_parquet(batches).map(Body::from)
+            // Grab the first batch so that we can get the schema
+            let Some(batch) = stream.next().await.transpose()? else {
+                return Ok(Body::empty());
+            };
+            let schema = batch.schema();
+
+            let mut bytes = Vec::new();
+            let mem_pool = Arc::new(UnboundedMemoryPool::default());
+            let mut writer = TrackedMemoryArrowWriter::try_new(&mut bytes, schema, mem_pool)?;
+
+            // Write the first batch we got and then continue writing batches
+            writer.write(batch)?;
+            while let Some(batch) = stream.next().await.transpose()? {
+                writer.write(batch)?;
+            }
+            writer.close()?;
+            Ok(Body::from(Bytes::from(bytes)))
         }
         QueryFormat::Csv => {
-            let batches = stream.try_collect::<Vec<RecordBatch>>().await?;
-            to_csv(batches).map(Body::from)
+            struct CsvFuture {
+                first_poll: bool,
+                stream: Pin<Box<dyn RecordBatchStream + Send>>,
+            }
+
+            impl Future for CsvFuture {
+                type Output = Option<Result<Bytes, DataFusionError>>;
+
+                fn poll(
+                    mut self: Pin<&mut Self>,
+                    ctx: &mut std::task::Context<'_>,
+                ) -> Poll<Self::Output> {
+                    match self.stream.poll_next_unpin(ctx) {
+                        Poll::Ready(Some(batch)) => {
+                            let batch = match batch {
+                                Ok(batch) => batch,
+                                Err(e) => return Poll::Ready(Some(Err(e))),
+                            };
+                            let mut writer = if self.first_poll {
+                                self.first_poll = false;
+                                arrow_csv::Writer::new(Vec::new())
+                            } else {
+                                arrow_csv::WriterBuilder::new()
+                                    .with_header(false)
+                                    .build(Vec::new())
+                            };
+                            writer.write(&batch).unwrap();
```

Comment:
Prior to this PR and still in the fully-buffered branches of the format match statement it looks like ` writer.write(&batch)?` is being called -- is it not possible to do that here because of the `Output` type of this future?

Reply:
Ah yeah, but I think it would need `.into()` called: https://docs.rs/datafusion/latest/datafusion/error/enum.DataFusionError.html#impl-From%3CError%3E-for-DataFusionError-1, right? I was initially assuming there was no `impl From<std::io::Error> for DatafusionError` and that it would need to be convered to the `Error` defined in this module.

Reply:
This ended up working out well with the into.

As for the `writer.write(&batch)` we can't use the ? anymore because you now need to return a `Poll` type. `async` is essentially doing all this under the hood, but since we're writing the future ourselves by hand we have this problem where we need to hand wrap these errors to work.

There is no From impl to make ? work for Poll as far as I'm aware.